### PR TITLE
Feature 1098 introduce data section in config file

### LINKED ISSUE #1098: Introduce data section inside sechub configuration file
## Situation
We currently provide only source code uploads inside `codeScan `configuration areas.
But we have planned to support also
- secretScan
- codeScan (with binaries)
- dependencyScan (with sources and binaries)
- licenseScan
- ... more ...

Which contains binary content as well - and this is currently not possible inside the configuration.
Also it is not possible to reuse an upload for all kind of scan types (e.g. when defining a secretScan and a codeScan in same configuration file )

## Wanted
As a SecHub user I want to define a configuration file where I define my data _(binary and/or source content)_, upload it once and use this for different kind of scans in ONE job.

Existing job configurations for users currently only performing code scans shall still work.

### Codescan can be defined as before
We will provide following as before:
```json
{
      "apiVersion": "1.0",
      "server"    : "https://sechub.example.org",

      "project"   : "gamechanger",

      "codeScan": {
         "fileSystem": {
              "folders": ["gamechanger-android/src/main/java",
                          "gamechanger-server/src/main/java"]
         },
         "excludes": ["**/mytestcode/**","**/documentation/**"],
         "additionalFilenameExtensions": [".cplusplus",".py9"]
      }
}
```
But the segment for code scan ("fileSystem" ) is only a shortcut for:

```json
{
      "apiVersion": "1.0",
      "server"    : "https://sechub.example.org",

      "project"   : "gamechanger",
      
      "data" : {
           "sources": [
             {
                "fileSystem": { 
                    "folders": ["gamechanger-android/src/main/java",
                            "gamechanger-server/src/main/java"] 
                 },
              "excludes": ["**/mytestcode/**","**/documentation/**"],  
              "additionalFilenameExtensions": [".cplusplus",".py9"]  
            }
          ]
      },

      "codeScan": { 
         
      }
}
```
### New possiblities for multiple scans for same upload or binary content

While introducing "data" as new element inside our json configuration, we will also
- provide **mandatory** `name` attribute inside data object elements
- we introduce also `files` inside `fileSystem` so we can define also single files
- we can reference named object elements via "use" attribute inside different scan type configurations 

Here a full blown example (with massive configurations):

```json
{
  "apiVersion" : "1.0",
  "server" : "https://sechub.example.org",
  "project" : "gamechanger",
  "data" : { // contains the different kind of data objects ( sources, binaries)
    "sources" : [ {
      "name" : "android-sources",
      "fileSystem" : {
        "folders" : [ "gamechanger-android/src/main/java", "gamechanger-server/src/main/java" ]
      },
      "excludes" : [ "**/mytestcode/**", "**/documentation/**" ],
      "additionalFilenameExtensions" : [ ".cplusplus", ".py9" ]
    }, {
      "name" : "openapi-reference",
      "fileSystem" : {
        "files" : [ "gamechanger-webapp/src/main/resources/openapi3.json" ]
      }
    }, {
      "name" : "secret-scanner-config",
      "fileSystem" : {
        "files" : [ "gamechanger-data/src/main/resources/secret-scanner-dictionary.json" ] // new: files
      }
    } ],
    "binaries" : [ { // new: binaries!
      "name" : "android-libraries",
      "fileSystem" : {
        "folders" : [ "gamechanger-android/lib/dlls", "gamechanger-server/lib/so" ]
      },
      "excludes" : [ "**/*.jpg/**", "myimages/**" ]
    } ]
  },
  // Define wanted scan and reference - if wanted the named data object
  "codeScan" : {
    "use" : [ "android-sources","android-libraries","openapi-reference"]
  },
  "dependencyScan" : { // fictional: currently this scan type does not exist, just an example
    "use" : [ "android-sources", "android-libraries" ]
  },
  "secretScan" : {// fictional: currently this scan type does not exist, just an example
    "use" : [ "secret-scanner-config" ],
    "featureToggle" : {
      "entropy" : "off"
    }
  },
  "licenseScan" : {// fictional: currently this scan type does not exist, just an example
    "use" : [ "android-sources","android-libraries" ]
  },
  "webScan" : {
    "url" : "https://example.com",
    "openApi" : {
      "use" : [ "openapi-reference"] 
    }
  }
}
```

### LINKED ISSUE #1173: Target resolving shall be done only for webscans and infrascans
## Situation

While defining and implementing
- #1098 
- #1163 
- #1164
It became clear that, with the introduction of the data-section, the concept of target resolving would lead to a massive implementation effort:
  - binary uploads  
  - data section parts at all
  - defining how the filter mechanism would apply to the targets...
  - .. much more. ..
Without any real purpose (the former code scan target resolving was only done, because it was designed for every scan)

Also the #1164 will filter files by configuration and there is no need to filter targets for source code manually

## Wanted
Target resolving shall only be done where necessary (to differ between INTRANET and INTERNET) but not for all

## Solution
- Change implementation, so target resolving is only done where necessary
- Reduce complexity, will be done with #1213

### LINKED ISSUE #1177: Problems in IDE with duplicated actuator
## Situation
When I start SecHub server inside my IDE (eclipse) with

```
-Dspring.profiles.active=mocked_products,h2,integrationtest
-Dsechub.server.debug=false
-Dsechub.storage.sharedvolume.upload.dir=temp
-Dsechub.targettype.detection.intranet.hostname.endswith=intranet.example.org
-Dsechub.config.trigger.nextjob.initialdelay=0
-Dsechub.storage.sharedvolume.upload.dir=/home/username/sechub-data/upload
-Dsechub.development.webdesignmode.enabled=true

```
And executing `JobScenario2IntTest` I got an Illegal argument exception when second batch job is executed:
```
Failed to register Collector of type MicrometerCollector: 
The Collector exposes the same name multiple times: spring_batch_job_active_seconds
```
But when integration tests were executed by github actions via gradle the tests did NOT fail.
**So gradle and eclipse IDE run do differ here!**

## Environment
I tested with
- Commit 7e8611ad from branch `develop`
  (Spring Boot 2.6.4)
- Commit 26e197b2 from branch `feature-1098-...`
- Java 8, Java 11
- Eclipse 2021-12
- Eclipse 2022-03
- Eclipse project setup done by gradle call
  `./gradlew cleanEclipse clipse` and a refresh inside eclipse IDE
- Gradle 7.4
- Linux

Always the first test works, after the job has been created, all other tests do fail, because on server side the mentioned exception happens. 
After this it's no longer possible to execute any job test from IDE without restarting the integration test server launch again.

## Wanted
Shall work again in eclipse - this is big problem and slows down development.

### When? 
- After every job this happens
   After the first job has been processed, the problem appears on server side. Means even on first (succesful) test!

### How is the server started in eclipse?
I inspected the properties of the server launch in eclipse  (for commit 26e197b2)  ... and I found something very odd (I added some new lines here and it's not the full java command, but it shows the problem):
image

There are two different micrometer-prometheus jars inside the classpath:
```
micrometer-registry-prometheus-1.8.3.jar
micrometer-registry-prometheus-1.7.2.jar
```

Looking at the depdencies of `sechub-server` I found:
image

So inside gradle the library version is updated from 1.7.2 to 1.8.3 (and so should only use 1.8.3)

But where comes the 1.7.2?
Reason found:
image
****

## Solution
The problem is, that for eclipse .classpath generation the gradle eclipse plugin does use not the commonly used jar version, but the project defined one.

I am not sure if 
Change `build.gradle` part to `implementation 'io.micrometer:micrometer-registry-prometheus'

### LINKED ISSUE #1213: Reduce complexity of product executors and target handling
## Situation 
Currently there are many install setup classes which are used inside different abstract product executor classes in combination with generics and inheritance etc. etc.

It is very confusing to read the code, maintain it, understand what is going on etc. (even for my self and I wrote it...)

Also the effort to keep the implementation for #1098  are too big, so it is time to cleanup the legacy stuff.

## Wanted
A simpler class model, less inheritance, less generics, more clear class names

## Solution
Remark: The commit f7d19111 handled this but did contain wrong issue number.

- [x]  Reduce complexity 
  - InstallSetup 
   It is so much complex with all the install setup classes, abstract base classes, abstract base product executor classes and so on. The credo is: "composition over inheritance" ... but was not followed (by myself...). So must be refactored!
  - Product Executors
    Simplify to ONE abstract base class only, drop other abstract classes
  - Remove additional Product Executor interfaces
    - Instead use the scan type only,
    - Introduce ScanType.REPORT to handle reporting product executors
    - Simplify and adopt former changes for report executor collection inside product executor services, 
- Drop unnecessary tests
- Drop TargetType `CodeScan` completely
- Will be done together with #1098 
- Renaming
  - We will also rename `Target` to `NetworkTarget` (and related parts as well)
  - `*NetworkTargetData` will be renamed to `*NetworkTargetProductServerData`

- [x]  Document changes

### PR DESCRIPTION
**Pull request by de-jcup**
This PR
- closes #1098  
- closes #1177
- closes #1173
- closes #1213

### CODE REVIEW

File: sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/NetworkTargetProductServerDataProvider.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+    String getUsernameWhenInternetTarget();
+
+    String getUsernameWhenIntranetTarget();
+
+    String getPasswordWhenInternetTarget();
+
+    String getPasswordWhenIntranetTarget();
+
+    boolean isHavingUntrustedCertificateForIntranet();
```
Comment (Jeeppler) [inline sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/NetworkTargetProductServerDataProvider.java:22]: `isHaving` is weird. Please, use `hasUntrusted… `

File: sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/NetworkTargetProductServerDataProvider.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    boolean isHavingUntrustedCertificateForInternet();
```
Comment (Jeeppler) [inline sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/NetworkTargetProductServerDataProvider.java:24]:
```suggestion
    boolean hasUntrustedCertificateForInternet();
```

File: sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/admin/FullScanDataRestController.java [hunk 3]
```diff
@@ -62,7 +62,7 @@ public void getFullScanZipFileForJob(
             support.writeScanData(fullScanData, outputStream);
         } catch (IOException e) {
             LOG.error("Was not able to provide zip file for full scan data of {}", logSanitizer.sanitize(sechubJobUUID, -1), e);
-            throw new NotFoundException("Was not able to support zip file, see logs for details");
+            throw new NotFoundException("Was not able to networkTargetDataSupport zip file, see logs for details");
```
Comment (Jeeppler) [inline sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/admin/FullScanDataRestController.java:65]: What do you mean? What does `networkTargetDataSupport` and `zip file` have in common?

File: sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/AbstractProductExecutor.java [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+        configureNetworkTargetHandlingIfNecessary(data);
+
+        return startExecution(data);
+    }
+
+    private void configureSourceCodeHandlingIfNecessary(ProductExecutorData data) {
+        if (scanType != ScanType.CODE_SCAN) {
+            return;
+        }
+        // the information about pathes is interesting for debugging but also necessary
```
Comment (Jeeppler) [inline sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/AbstractProductExecutor.java:117]:
```suggestion
        // the information about paths is interesting for debugging but also necessary
```

File: sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/AbstractProductExecutor.java [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+        // for our integration tests - see mocked_setup.json
+        Set<String> pathes = new LinkedHashSet<>();
```
Comment (Jeeppler) [inline sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/AbstractProductExecutor.java:119]:
```suggestion
        Set<String> paths = new LinkedHashSet<>();
```

File: sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/AbstractProductExecutor.java [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+            if (!(config instanceof SecHubSourceDataConfiguration)) {
+                LOG.warn("source object data was not expected {} but {}", SecHubSourceDataConfiguration.class, config.getClass());
+                continue;
+            }
+            SecHubSourceDataConfiguration sourceDataConfig = (SecHubSourceDataConfiguration) config;
+            addFileSystemParts(pathes, sourceDataConfig);
+        }
+    }
+
+    private void addFileSystemParts(Set<String> pathes, SecHubFileSystemContainer container) {
```
Comment (Jeeppler) [inline sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/AbstractProductExecutor.java:148]:
```suggestion
    private void addFileSystemParts(Set<String> paths, SecHubFileSystemContainer container) {
```

File: sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/AbstractProductExecutor.java [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+    }
+
+    private void configureNetworkTargetHandlingIfNecessary(ProductExecutorData data) {
+        if (!isUsingNetworkTargets()) {
+            return;
+        }
+        /* check preconditions */
+        NetworkTargetProductServerDataProvider networkTargetDataProvider = data.networkTargetDataProvider;
+        if (networkTargetDataProvider == null) {
+            throw new IllegalStateException("No network target data provider set, but necessary for scantype:" + scanType
```
Comment (Jeeppler) [inline sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/AbstractProductExecutor.java:207]:
```suggestion
            throw new IllegalStateException("No network target data provider set, but necessary for scantype: " + scanType
```

File: sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/AbstractProductExecutor.java [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+                    + "\nInject this at customize method inside " + getClass().getName());
+        }
+
+        NetworkLocationProvider networkLocationProvider = data.networkLocationProvider;
+        if (networkLocationProvider == null) {
+            throw new IllegalStateException("No network location provier set, but necessary for scantype:" + scanType
```
Comment (Jeeppler) [inline sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/AbstractProductExecutor.java:213]:
```suggestion
            throw new IllegalStateException("No network location provier set, but necessary for scantype: " + scanType
```

**Review by Jeeppler — CHANGES_REQUESTED** [review]
## Review part 1

In general, the code is easier to read and cleaner. However, I found a couple mistakes. Have a look at the code.

**Review by Jeeppler — CHANGES_REQUESTED** [review]
## Review Part 2

See comments.

File: sechub-commons-model/src/test/java/com/mercedesbenz/sechub/commons/model/SecHubConfigurationModelValidatorTest.java [hunk 9]
```diff
@@ -0,0 +1,523 @@
...
+        /* test */
+        assertFalse(result.hasErrors());
+
+    }
+
+    @Test
+    void infrascanconfig_with_one_ip_results_in_no_error() throws Exception {
+        /* prepare */
+        SecHubInfrastructureScanConfiguration infraScan = new SecHubInfrastructureScanConfiguration();
+        infraScan.getIps().add(mock(InetAddress.class));// &getByName("localhost"));
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-commons-model/src/test/java/com/mercedesbenz/sechub/commons/model/SecHubConfigurationModelValidatorTest.java:120]: Old stuff from former test - was out commented... and forgotten. Will remove this

File: sechub-commons-model/src/test/java/com/mercedesbenz/sechub/commons/model/SecHubConfigurationModelValidatorTest.java [hunk 10]
```diff
@@ -0,0 +1,523 @@
...
+    }
+
+    @Test
+    void model_having_one_elements_with_name_with_length101_has_error() {
+        /* prepare */
+        SecHubConfigurationModel model = createDefaultValidModel();
+        SecHubDataConfiguration data = new SecHubDataConfiguration();
+
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < 101; i++) {
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-commons-model/src/test/java/com/mercedesbenz/sechub/commons/model/SecHubConfigurationModelValidatorTest.java:301]: Thank you for the hint. But currently we still provide officially still Java 8 so I cannot do this. When we officially switch to Java >=11 we could change those parts. 
But it's just a test ...

File: sechub-scan-product-nessus/src/main/java/com/mercedesbenz/sechub/domain/scan/product/nessus/NessusInstallSetupImpl.java [hunk 11]
```diff
@@ -54,32 +52,47 @@ public String getDefaultPolicyId() {
...
+        return null;
+    }
+
+    @Override
+    public boolean isHavingUntrustedCertificateForIntranet() {
+        return trustAllCertificatesNecessary;
+    }
+
+    @Override
+    public boolean isHavingUntrustedCertificateForInternet() {
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-scan-product-nessus/src/main/java/com/mercedesbenz/sechub/domain/scan/product/nessus/NessusInstallSetupImpl.java:95]: I aggree that "has*" is more readable than "isHaving" - even when the Java Bean "contract" is not fullfilled (only setters/getters and is allowed). But this interface is not really a java bean - so has is better indeed.

This is an interface method.
Reading all other methods of `NetworkTargetProductServerDataProvider` I see that every other method uses at the end either "*WhenInternetTarget" or "*WhenInranetTarget"

So I will change
- isHavingUntrustedCertificateForInternet
- isHavingUntrustedCertificateForIntranet

to
- hasUntrustedCertificateForWhenInternetTarget
- hasUntrustedCertificateWhenIntranetTarget

File: sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/AbstractProductExecutor.java (hunk 4 above)
Comment (de-jcup) [reply sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/AbstractProductExecutor.java:117]: I will change your found parts here inside this PR. But there more than 300 of such findings and I do not want to touch too many other files.

So I decided to create https://github.com/mercedes-benz/sechub/issues/1228 for the additional ones.

File: sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/ProductExecutorData.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+    ProductExecutorContext productExecutorContext;
+    SecHubExecutionContext sechubExecutionContext;
+    UUIDTraceLogID traceLogId;
+    NetworkTargetProductServerDataSuppport networkTargetDataSupport;
+
+    NetworkLocationProvider networkLocationProvider;
+    NetworkTargetInfo currentNetworkTargetInfo;
+    String traceLogIdAsString;
+    Set<String> codeUploadFileSytemFolderPathes;
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/ProductExecutorData.java:27]: Found another typo... Sytem...

Will change to `codeUploadFileSystemFolderPaths`

File: sechub-scan/src/test/java/com/mercedesbenz/sechub/domain/scan/TargetTypeTest.java [hunk 13]
```diff
@@ -10,61 +10,47 @@
     @Test
     public void healthcheck() {
 
-        assertEquals("Hmm.. seems there is a new target typeToTest which is not tested, please add + modify...", 5, TargetType.values().length);
+        assertEquals("Hmm.. amount of available target types differs! eems there is a new target typeToTest which is not tested, please add + modify...", 4,
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-scan/src/test/java/com/mercedesbenz/sechub/domain/scan/TargetTypeTest.java:13]: Okay, but will also rename test to `NetworkTargetTypeTest`

File: sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/admin/FullScanDataRestController.java (hunk 3 above)
Comment (de-jcup) [reply sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/admin/FullScanDataRestController.java:65]: This was accidently changed by wrong refactoring setup in IDE ("change all occurrences + renamed "support" field  to "networkTargetDataSupport" .

This is not the only occurrence!

I will change this for all accidently renamed parts.

File: sechub-commons-model/src/test/java/com/mercedesbenz/sechub/commons/model/SecHubConfigurationModelValidatorTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (Jeeppler) [reply sechub-commons-model/src/test/java/com/mercedesbenz/sechub/commons/model/SecHubConfigurationModelValidatorTest.java:301]: I would like you to extract this snippet and place it in a small helper method. Repeating strings is a useful functionality.

We use the same idom (pettern) in:
- ProjectDescriptionValidationImpl
- ProfileDescriptionValidationImpl
- OneTimeTokenValidationImplTest

we have to re-implement this every time. As a result, it makes sense to use a helper method.

**Review by Jeeppler — APPROVED** [review]
Looks good to me.

**Pull request merged**